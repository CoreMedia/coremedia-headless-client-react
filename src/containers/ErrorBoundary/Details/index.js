// @flow
import styled from 'styled-components';

const Details = styled.details`
  white-space: pre-wrap;
`;
Details.displayName = 'Details';

export default Details;
