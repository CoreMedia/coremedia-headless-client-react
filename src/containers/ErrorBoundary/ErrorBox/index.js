// @flow
import { TeaserBox } from '../../../components/basic/Box';
import { getColors } from '../../../styles/themes/utils';

const ErrorBox = TeaserBox.extend`
  border: 2px solid ${props => getColors(props.theme.colors, 'red').bgsolid};
`;

ErrorBox.displayName = 'ErrorBox';

ErrorBox.defaultProps = {
  theme: {
    colors: {
      red: {
        bgsolid: 'rgba(221, 52, 40, 1)',
      },
    },
  },
};

export default ErrorBox;
