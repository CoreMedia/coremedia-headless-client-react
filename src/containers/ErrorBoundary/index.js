// @flow
import * as React from 'react';
import PropTypes from 'prop-types';

import ErrorBox from './ErrorBox';
import Details from './Details';
import { H3 } from '../../components/basic/Heading';

type Props = {
  children: React.Node,
};

type State = {
  error: ?Error,
  errorInfo: ?Object,
};

class ErrorBoundary extends React.Component<Props, State> {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  state = {
    error: null,
    errorInfo: null,
  };

  componentDidCatch(error: Error, errorInfo: Object) {
    this.setState(prevState => ({
      error: error,
      errorInfo: errorInfo,
    }));
  }

  render() {
    if (this.state.errorInfo) {
      return (
        <ErrorBox>
          <H3>Something went wrong...</H3>
          <Details>
            {this.state.error && this.state.error.toString()}
            <br />
            {this.state.errorInfo.componentStack}
          </Details>
        </ErrorBox>
      );
    }
    return this.props.children;
  }
}

export default ErrorBoundary;
