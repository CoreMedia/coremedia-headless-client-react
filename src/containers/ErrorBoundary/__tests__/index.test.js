// @flow
import React from 'react';

import ErrorBoundary from '..';

describe('ErrorBoundary Container', () => {
  it('should render children', () => {
    const wrapper = shallow(<ErrorBoundary>Content</ErrorBoundary>);
    expect(wrapper.state('errorInfo')).toEqual(null);
    expect(wrapper.state('error')).toEqual(null);
    expect(wrapper.text()).toEqual('Content');
  });

  it('should render ErrorBox Component', () => {
    const error = new Error('I am an error message.');
    const errorInfo = {
      componentStack: 'I am a componentStack.',
    };
    const wrapper = shallow(<ErrorBoundary>Content</ErrorBoundary>);
    wrapper.instance().componentDidCatch(error, errorInfo);
    expect(wrapper.state('error')).toEqual(error);
    expect(wrapper.state('errorInfo')).toEqual(errorInfo);
    expect(wrapper.update()).toMatchSnapshot();
  });
});
