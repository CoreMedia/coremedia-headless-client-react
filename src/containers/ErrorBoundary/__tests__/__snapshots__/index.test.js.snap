// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Container should render ErrorBox Component 1`] = `
<ErrorBox
  theme={
    Object {
      "colors": Object {
        "red": Object {
          "bgsolid": "rgba(221, 52, 40, 1)",
        },
      },
    }
  }
>
  <H3
    theme={
      Object {
        "breakpoints": Object {
          "desktop": 992,
          "large": 1280,
          "tablet": 768,
        },
        "name": "default",
      }
    }
  >
    Something went wrong...
  </H3>
  <Details>
    Error: I am an error message.
    <br />
    I am a componentStack.
  </Details>
</ErrorBox>
`;
